GoogleContainerTools/skaffold
# add debug logs and don't return on first terminated state

Fixes #4589

In this PR
In `daig` packages, 
- Don't return on first successfully terminated container to continue going through waiting container statuses.
   If any of the init containers terminated successfully, Pod Validator would stop going through rest of the containers.
   Hence we see the issue in #4589 
- Add some debug statements

In skaffold, status-check
- @dgageot addresses the concern in https://github.com/GoogleContainerTools/skaffold/issues/4589#issuecomment-665878771  where status check end will now stop polling for status if a pod container can't recover unless a code change is made. 
Created #4599 


To verify:
1. clone https://github.com/vincentjocodes/nextjs-test-app-cloud-code
2. Change the last line in [Dockerfile](https://github.com/vincentjocodes/nextjs-test-app-cloud-code/blob/master/Dockerfile)
```
ENTRYPOINT ["node", "./node_modules/.bin/next", "dev", "-p", "8080", "--inspect=127.0.0.1:9229"]
```
3. run `skaffold debug` from the executable built on this branch
```
~/workspace/skaffold/out/skaffold debug -d <>
```
4. Verify container logs are seen in the status check logs
```
Tags used in deployment:
 - test-app -> gcr.io/tejal-test/test-app:latest@sha256:79abb366cb3dba0e05a60f1f4e176f59d08c56473e9f6438675f5618629466b1
Starting deploy...
 - deployment.apps/test-app configured
 - service/test-app-external configured
Waiting for deployments to stabilize...
 - deployment/test-app: waiting for rollout to finish: 1 old replicas are pending termination...
 - deployment/test-app: waiting for rollout to finish: 1 old replicas are pending termination...
 - deployment/test-app: waiting for rollout to finish: 1 old replicas are pending termination...
    - pod/test-app-7ff8788454-vf8vl: container server terminated with exit code 1
[test-app-7ff8788454-vf8vl server] /src/app/node_modules/next/dist/bin/next:25
[test-app-7ff8788454-vf8vl server]   `);process.exit(0);}const command=foundCommand?args._[0]:defaultCommand;const forwardedArgs=foundCommand?args._.slice(1):args._;if(args['--inspect'])throw new Error(`--inspect flag is deprecated. Use env variable NODE_OPTIONS instead: NODE_OPTIONS='--inspect' next ${command}`);// Make sure the `next <subcommand> --help` case is covered
[test-app-7ff8788454-vf8vl server]                                                                                                                                                        ^
[test-app-7ff8788454-vf8vl server] 
[test-app-7ff8788454-vf8vl server] Error: --inspect flag is deprecated. Use env variable NODE_OPTIONS instead: NODE_OPTIONS='--inspect' next dev
[test-app-7ff8788454-vf8vl server]     at Object.<anonymous> (/src/app/node_modules/next/dist/bin/next:25:158)
[test-app-7ff8788454-vf8vl server]     at Module._compile (internal/modules/cjs/loader.js:688:30)
[test-app-7ff8788454-vf8vl server]     at Object.Module._extensions..js (internal/modules/cjs/loader.js:699:10)
[test-app-7ff8788454-vf8vl server]     at Module.load (internal/modules/cjs/loader.js:598:32)
[test-app-7ff8788454-vf8vl server]     at tryModuleLoad (internal/modules/cjs/loader.js:537:12)
[test-app-7ff8788454-vf8vl server]     at Function.Module._load (internal/modules/cjs/loader.js:529:3)
[test-app-7ff8788454-vf8vl server]     at Function.Module.runMain (internal/modules/cjs/loader.js:741:12)
[test-app-7ff8788454-vf8vl server]     at startup (internal/bootstrap/node.js:285:19)
[test-app-7ff8788454-vf8vl server]     at bootstrapNodeJSCore (internal/bootstrap/node.js:739:3)
[test-app-7ff8788454-vf8vl server] time="2020-07-29T22:13:27Z" level=fatal msg="exit status 1"
```

**Make sure, the error message is appears once.**

To verify status check is interrupted in dev loop.
1. `skaffold dev` on a running example
2. make a change in `deployment.yaml` which is results in invalid pod config 
    e.g. Deployment.Spec.Pod.Image -> points to invalid image name.
2. Verify skaffold stops status-check phase and starts watching for file changes.
